src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -198,6 +198,12 @@ private async Task WriteToOutputPipe()
                     // Now check the connection window
                     actual = CheckConnectionWindow(actual);
 
+                    // reset to zero, if negative
+                    if (actual < 0)
```

Comment:
This does seem to fix it; it *feels* like we're powering over the real problem, though - like one of the values used in either CheckStreamWindow or CheckConnctionWindow is already at a value we should have clamped. Maybe I'm overthinking... I tend to do that, and this has the advantage of simplicity

Reply:
I vaguely recall that the spec allows the window size to go negative.  I do think it might be helpful to include a comment about when that can happen.

Reply:
Do we actually want to continue if it's negative/zero or should we short-circuit something?

Reply:
@amcasey yes, when the window size becomes we (sender) should wait for window update from the receiver, by continuing the code flow will mark as waiting for window updates, as there are still more data to be written. This will also handle the case if the receiver aborts the connection without sending window update.
Added inline comment to describe the change

Reply:
Crucially, we're only setting a local to 0, right?  Not our actual window size state?

Reply:
yes, that's correct, only the local is set to 0
the actual window size remains negative